docs/new-architecture-playbook.md

```diff
@@ -0,0 +1,2182 @@
+
+Update your native module or component to ensure it **extends the abstract class** that has been code-generated from your JavaScript specs (i.e. the `NativeAwesomeManagerSpec.java` file from the previous example).
+
+Following the example set forth in the previous section, your library might import `NativeAwesomeManagerSpec`, implement the relevant native interface and the necessary methods for it:
+
+```java
+import androidx.annotation.NonNull;
+
+import com.example.samplelibrary.NativeAwesomeManagerSpec;
+import com.facebook.react.bridge.Promise;
+import com.facebook.react.bridge.ReactApplicationContext;
+
+public class NativeAwesomeManager extends NativeAwesomeManagerSpec {
+
+    public static final String NAME = "NativeAwesomeManager";
+
+    public NativeAwesomeManager(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @Override
+    public void getString(String id, Promise promise) {
+        // Implement this method
+    }
+
+    @NonNull
+    @Override
+    public String getName() {
+        return NAME;
+    }
+}
+```
+
+Please note that the **generated abstract class** that you’re now extending (`MyAwesomeSpec` in this example), is itself extending `ReactContextBaseJavaModule`. Therefore you should not use access to any of the method/fields you were previously using (e.g. the `ReactApplicationContext` and so on). Moreover the generated class will now also implement the `TurboModule` interface for you.
+
+## Enabling the New Architecture in your Library (iOS)
+
+You have defined the JavaScript specs for your native modules as part of the [prerequisites](#turbomodules-define-specs-in-javascript) and you are now ready to migrate your library to the new architecture. Here are the steps you can follow to accomplish this.
+
+### 1. Updating your Podspec for the new architecture
+
+The new architecture makes use of CocoaPods.
+
+#### Add Folly and Other Dependencies
+
+We'll need to ensure Folly is configured properly in any projects that consume your library. With CocoaPods, we can use the `compiler_flags` and `dependency` properties to set it up.
+
+Add these to your `Pod::Spec.new` block:
+
+```ruby
+# folly_version must match the version used in React Native
+# See folly_version in react-native/React/FBReactNativeSpec/FBReactNativeSpec.podspec
+folly_version = '2021.06.28.00-v2'
+folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32'
+
+Pod::Spec.new do |s|
+  # ...
+  s.compiler_flags  = folly_compiler_flags
+
+  s.pod_target_xcconfig    = {
+    "HEADER_SEARCH_PATHS" => "\"$(PODS_ROOT)/boost\""
+  }
+
+  s.dependency "React"
+  s.dependency "React-RCTFabric" # This is for fabric component
+  s.dependency "React-Codegen"
+  s.dependency "RCT-Folly", folly_version
+  s.dependency "RCTRequired"
+  s.dependency "RCTTypeSafety"
+  s.dependency "ReactCommon/turbomodule/core"
+  # ...
+end
+```
```

Comment:
I think this will break Swift Modules, since a few of those Podspecs (e.g. `ReactCommon/turbomodule/core`) do not expose Swift Modules / Modular Headers. This will prevent people from using Swift.

While working on VisionCamera, I've made a PR to fix this in the react-native core repo: https://github.com/facebook/react-native/pull/31858 - but that was a long time ago, not sure if that still builds.

Reply:
cc @sota000
I'm not really familiar with Swift support. Are you suggesting that we add `"DEFINES_MODULE" => "YES"` to `ReactCommon/turbomodule/core` and the other modules this pod depends on?

Reply:
Yes I think that is the required flag for this. I think those are the three most important points for a Podspec to allow it being used in Swift modules:
* Have `DEFINES_MODULE` set to `YES`
* Ignore all headers by default using `preserve_paths`, because those might be C++ headers and importing them in Swift will break. This will also prevent nameclashes on building, see https://github.com/firebase/firebase-ios-sdk/issues/4035 for more details on that.
  - For those headers that are plain Objective-C headers and don't contain C++, we can add them to `s.source_files`. This will make them importable by other modules, including Swift.
  - For those headers that are Objective-C++ (or C++) and still need to be exported to be importable by other C++ (non-Swift) modules, we can either create a Subspec specifically for those, or wrap them with `#ifdef __cplusplus` as a last resort?
* `s.requires_arc = true` (I think)

I think there are some improvements on Swift's C++ interoperability (see [discussion here](https://forums.swift.org/t/swift-and-c-interoperability-workgroup-announcement/54998)), but I'm not sure how advanced that is (those efforts started 3 years ago)